ext/MatrixAlgebraKitMooncakeExt/MatrixAlgebraKitMooncakeExt.jl

```diff
@@ -0,0 +1,352 @@
+            # of ComplexF32) into the correct **forwards** data type (since we are now in the forward
+            # pass). For many types this is done automatically when the forward step returns, but
+            # not for nested structs with various fields (like Diagonal{Complex})
+            USVᴴ_codual = Mooncake.CoDual(USVᴴ, Mooncake.fdata(Mooncake.zero_tangent(USVᴴ)))
+            function dsvd_adjoint(::Mooncake.NoRData)
+                U, S, Vᴴ = Mooncake.primal(USVᴴ_codual)
+                dU_, dS_, dVᴴ_ = Mooncake.tangent(USVᴴ_codual)
+                U, dU = Mooncake.arrayify(U, dU_)
+                S, dS = Mooncake.arrayify(S, dS_)
+                Vᴴ, dVᴴ = Mooncake.arrayify(Vᴴ, dVᴴ_)
+                if $(f == svd_compact)
+                    svd_pullback!(dA, A, (U, S, Vᴴ), (dU, dS, dVᴴ))
+                else # full
+                    minmn = min(size(A)...)
+                    vU = view(U, :, 1:minmn)
+                    vS = Diagonal(view(diagview(S), 1:minmn))
+                    vVᴴ = view(Vᴴ, 1:minmn, :)
+                    vdU = view(dU, :, 1:minmn)
+                    vdS = Diagonal(view(diagview(dS), 1:minmn))
+                    vdVᴴ = view(dVᴴ, 1:minmn, :)
+                    svd_pullback!(dA, A, (vU, vS, vVᴴ), (vdU, vdS, vdVᴴ))
+                end
+                MatrixAlgebraKit.zero!(dU)
+                MatrixAlgebraKit.zero!(dS)
+                MatrixAlgebraKit.zero!(dVᴴ)
+                return Mooncake.NoRData(), Mooncake.NoRData(), Mooncake.NoRData()
+            end
+            return USVᴴ_codual, dsvd_adjoint
+        end
+    end
+end
+
+@is_primitive Mooncake.DefaultCtx Mooncake.ReverseMode Tuple{typeof(MatrixAlgebraKit.svd_vals!), Any, Any, MatrixAlgebraKit.AbstractAlgorithm}
+function Mooncake.rrule!!(::CoDual{typeof(MatrixAlgebraKit.svd_vals!)}, A_dA::CoDual, S_dS::CoDual, alg_dalg::CoDual)
+    # compute primal
+    S_ = Mooncake.primal(S_dS)
+    dS_ = Mooncake.tangent(S_dS)
+    A_ = Mooncake.primal(A_dA)
+    dA_ = Mooncake.tangent(A_dA)
+    A, dA = arrayify(A_, dA_)
+    S, dS = arrayify(S_, dS_)
+    U, nS, Vᴴ = svd_compact(A, Mooncake.primal(alg_dalg))
+    copy!(S, diagview(nS))
+    function dsvd_vals_adjoint(::Mooncake.NoRData)
+        svd_pullback!(dA, A, (U, S, Vᴴ), (nothing, dS, nothing))
+        MatrixAlgebraKit.zero!(dS)
+        return Mooncake.NoRData(), Mooncake.NoRData(), Mooncake.NoRData(), Mooncake.NoRData()
+    end
+    return S_dS, dsvd_vals_adjoint
+end
+
+@is_primitive Mooncake.DefaultCtx Mooncake.ReverseMode Tuple{typeof(MatrixAlgebraKit.svd_vals), Any, MatrixAlgebraKit.AbstractAlgorithm}
+function Mooncake.rrule!!(::CoDual{typeof(MatrixAlgebraKit.svd_vals)}, A_dA::CoDual, alg_dalg::CoDual)
+    # compute primal
+    A = Mooncake.primal(A_dA)
+    dA_ = Mooncake.tangent(A_dA)
+    A, dA = arrayify(A, dA_)
+    S = svd_vals(A, Mooncake.primal(alg_dalg))
+    U, _, Vᴴ = svd_compact(A, Mooncake.primal(alg_dalg))
```

Comment:
Calling `svd_` twice could become a real overhead. Any reason not to simply use the following:

```suggestion
    U, S_, Vᴴ = svd_compact(A, Mooncake.primal(alg_dalg))
    S = diagview(S_)
```